Source/WebKit/UIProcess/Launcher/haiku/ProcessLauncherHaiku.cpp

```diff
@@ -26,142 +26,76 @@
-#include <assert.h>
+#include <spawn.h>
 #include <unistd.h>
 
 using namespace WebCore;
 
 namespace WebKit {
 
-// TODO we could have a single instance of this instead of creating and deleting one everytime.
-// But it would need a way to either locate the ProcessLauncher from the processIdentifier,
-// or to store the ProcessLauncher pointer into the message sent to the other process and back
-// into the reply
-//
-// Another option would be to have the BHandler already created by the connectionIdentifier
-// and then ownership transferred to the corresponding ConnectionHaiku when it is created, instead
-// of creating yet another BHandler there.
-class ProcessLauncherHandler: public BHandler
-{
-    public:
-        ProcessLauncherHandler(ProcessLauncher* launcher)
-            : BHandler("process launcher")
-            , m_launcher(launcher)
-        {
-        }
-
-        void MessageReceived(BMessage* message)
-        {
-            switch(message->what)
-            {
-                case 'inig':
-                    GlobalMessage(message);
-                    break;
-                default:
-                    BHandler::MessageReceived(message);
-                    break;
-            }
-        }
-
-    private:
-
-        void GlobalMessage(BMessage* message)
-        {
-            WTF::ProcessID processID = message->FindInt64("processID");
-            BMessenger messenger;
-            message->FindMessenger("messenger", &messenger);
-            IPC::Connection::Identifier connectionIdentifier(std::move(messenger));
-            connectionIdentifier.m_isCreatedFromMessage = true;
-            m_launcher->didFinishLaunchingProcess(processID, connectionIdentifier);
-
-            // Our job is done!
-            // FIXME or is it? maybe keep this around until the processLauncher is destroyed?
-            delete this;
-        }
-
-        ProcessLauncher* m_launcher;
-};
-
-static BHandler* GetProcessLauncherHandler(ProcessLauncher* launcher)
-{
-    BHandler* handle = nullptr;
-    handle = new ProcessLauncherHandler(launcher);
-    BLooper* looper = BLooper::LooperForThread(find_thread(NULL));
-    looper->AddHandler(handle);
-
-    return handle;
-}
-
-static status_t GetRefForPath(BString path, entry_ref* pathRef)
-{
-    BEntry pathEntry(path);
-    if(!pathEntry.Exists())
-        return B_BAD_VALUE;
-
-    status_t result = pathEntry.GetRef(pathRef);
-    if(result != B_OK)
-        return result;
-
-    return B_OK;
-}
-
 void ProcessLauncher::launchProcess()
 {
-    BString executablePath;
+    IPC::SocketPair socketPair = IPC::createPlatformConnection(
+        IPC::PlatformConnectionOptions::SetCloexecOnClient | IPC::PlatformConnectionOptions::SetCloexecOnServer);
 
+    char* executablePath;
     switch (m_launchOptions.processType) {
     case ProcessLauncher::ProcessType::Web:
-        executablePath = executablePathOfWebProcess();
+        executablePath = strdup(executablePathOfWebProcess().ascii().data());
```

Comment:
was there a problem with using BString here to track the memory ownership? It would be easier to maintain if we don't have to manually free the memory.